### PR TITLE
[feat]: [PL-30916]: Created Job to remove User Role assignments behind FF.

### PR DESCRIPTION
**Pull request by Jimit09Gandhi**
## Describe your changes

- Removing User assigned viewer role assignments across all scopes.
- This will help to customers manage role assignments using Default User Groups instead of individual user.
- We had introduced Default User Groups at all scopes with AccountViewer(Role) - All Account Level resources(Resource Group) role binding as deletable and there are customers who have removed the role binding from Default User Group.
- Also for accounts having ACCOUNT_BASIC_ROLE_ONLY FF ON, we had not assigned AccountViewer - All Account Level resources role binding to Account Scope user group.
- So for both of this scenarios we wont be removing Account Viewer role binding from the users.
- To avoid having one time bloating up OpLog and direct load on Access Control service for processing delete events I have added delay of 1 sec after processing every account.
- For Prod-2 having max no of Accounts of all clusters it will add delay of roughly ~2.5 hours. But note this be run only once with ~90% Accounts.

## Checklist
- [X] I've documented the changes in the PR description.
- [X] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)

### CODE REVIEW

**Comment by Jimit09Gandhi** [comment]
trigger messagecheck

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOOxsW8pRvjStAeEHX:-NMOOxsW8pRvjStAeEHY:b-sogssm) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L72)):*
> ```Java
>   @Override
>   public void run() {
>     log.info(DEBUG_MESSAGE + "started...");
> ```

Comment from @ankit-harness  :   We should take a lock so that only 1 Pod is executing the Job at any point in time. Check InstanceSyncPerpetualTaskMigrationJob
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPPR69SBWVNGpjX-r:-NMOPPR69SBWVNGpjX-s:b-5iltrb) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L72)):*
> ```Java
>   @Override
>   public void run() {
>     log.info(DEBUG_MESSAGE + "started...");
> ```

Comment from @ankit-harness  : Lets say an account has been migrated. Admin gives some user role assignment directly. now, would those role assignments be migrated again on next run of the job? if yes, i would prefer we maintain state about which accounts have been migrated.
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOP4gf33lQTnjTR7im:-NMOP4gf33lQTnjTR7in:b-8qt9gl) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L94)):*
> ```Java
>     }
>     List<AccountDTO> ngEnabledAccounts =
>         accountDTOS.stream().filter(AccountDTO::isNextGenEnabled).collect(Collectors.toList());
> ```

Comment from @ankit-harness  :  Shouldn't we do it for all the accounts? nextGenEnabled may have been turned off after turned on. We should cleanup those accounts as well.
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 101 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPHO-5EzKvHN22KH8:-NMOPHO-5EzKvHN22KH9:b-zh2q0) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L101)):*
> ```Java
>     }
>     try {
>       List<String> filteredAccountsByFF = filterAccountsForAccountBasicRoleOnlyFF(targetAccounts);
> ```

Comment from @ankit-harness  :This will bombard the ff service service. We should complete one account and then pick another rather than filtering initially. That will spread out the load
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java` line 40 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPckS2NznO1zfi--M:-NMOPckS2NznO1zfi--N:b-vg0mc5) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java#L40)):*
> ```Java
>   @Override
>   public void start() throws Exception {
>     log.info(DEBUG_MESSAGE + "started...");
> ```

Comment from @ankit-harness  : We should have a way to disable this migration. Basically if something goes wrong, we have a way to disable it via config and redeploy
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java` line 42 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPXMJ2T20ugCWSLbT:-NMOPXMJ2T20ugCWSLbU:bhql85s) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java#L42)):*
> ```Java
>     log.info(DEBUG_MESSAGE + "started...");
>     userRoleAssignmentRemovalJobFuture =
>         executorService.scheduleWithFixedDelay(userRoleAssignmentRemovalMigrationJob, 15, 1440, TimeUnit.MINUTES);
> ```

Comment from @ankit-harness  : We should add large random initial delay so that pods don't execute this job in succession.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java` line 40 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPckS2NznO1zfi--M:-NMOQJS60MXWLBaIpOB4:bx0x4fa) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalService.java#L40)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Comment from @ankit-harness  : We should have a way to disable this migration. Basically if something goes wrong, we have a way to disable it via config and redeploy
</blockquote></details>

@ankit-harness  We are already way to do using FF PL_REMOVE_USER_VIEWER_ROLE_ASSIGNMENTS which can globally disabled.
And since we are using FF SDK, it caches all FF in host service. Do we need additional config?

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPPR69SBWVNGpjX-r:-NMOWisZAWex65WUui8r:b-uvrdnp) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L72)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Comment from @ankit-harness  : Lets say an account has been migrated. Admin gives some user role assignment directly. now, would those role assignments be migrated again on next run of the job? if yes, i would prefer we maintain state about which accounts have been migrated.
</blockquote></details>

@ankit-harness we are deleting only managed role assignment. When we release this Job any new role assignment would be
created with managed= false and so those would not be deleted.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOOxsW8pRvjStAeEHX:-NMOsydxCm9EHLZxqcU8:bqu797s) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L72)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Comment from @ankit-harness  :   We should take a lock so that only 1 Pod is executing the Job at any point in time. Check InstanceSyncPerpetualTaskMigrationJob
</blockquote></details>

Thanks for this catch @ankit-harness . 
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 101 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPHO-5EzKvHN22KH8:-NMOtAC53_47Nvk6Qr3Q:b8ncbft) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L101)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Comment from @ankit-harness  :This will bombard the ff service service. We should complete one account and then pick another rather than filtering initially. That will spread out the load
</blockquote></details>

FeatureFlagService uses FF SDK and it caches FF. So it wont be call to FF service. 
https://developer.harness.io/docs/feature-flags/ff-sdks/sdk-overview/communication-sdks-harness-feature-flags#how-do-the-sdks-communicate-with-feature-flags

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 101 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOPHO-5EzKvHN22KH8:-NMOu-l961n5YVl_yLkC:b-b8w2gg) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L101)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

FeatureFlagService uses FF SDK and it caches FF. So it wont be call to FF service. 
https://developer.harness.io/docs/feature-flags/ff-sdks/sdk-overview/communication-sdks-harness-feature-flags#how-do-the-sdks-communicate-with-feature-flags
</blockquote></details>

So I think we can evaluate for all accounts. Let me know your thoughts.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOOxsW8pRvjStAeEHX:-NMPiwcX8bO81O8RH5KP:bs7wub9) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L72)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Thanks for this catch @ankit-harness . 
</blockquote></details>

Done .
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/42835#-NMOP4gf33lQTnjTR7im:-NMPirAd0UnxwpQ7xtri:b-newbcy) ([raw file](https://github.com/harness/harness-core/blob/9ce0df71c85a16dd6d96cd4d9b172455cfe119b7/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L94)):*
> Comment from @ankit-harness  :  Shouldn't we do it for all the accounts? nextGenEnabled may have been turned off after turned on. We should cleanup those accounts as well.
Changed to remove for all accounts.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 6 of 9 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835#-:-NMRoHiICefu1cInosCz:bnfp4nl)* status: 9 file(s) to review. 1 of 2 approvals obtained.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835#-:-NMRoM6q8dqv2_xCaqts:bnfp4nl)* status: 9 file(s) to review. 1 of 2 approvals obtained.

File: access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+      log.error(DEBUG_MESSAGE + "Failed to fetch all accounts", ex);
+    }
+    List<String> targetAccounts = filterAccountsForFFEnabled(accountDTOS);
+    if (isEmpty(targetAccounts)) {
+      return;
+    }
+    try {
+      List<String> filteredAccountsByFF = filterAccountsForAccountBasicRoleOnlyFF(targetAccounts);
+      HashSet<String> filteredAccountIds =
+          new HashSet<>(filterAccountsWhereDefaultUserGroupDoesntHaveViewerRoleAssignment(filteredAccountsByFF));
```
Comment (meenakshi-harness) [inline access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java:116]: change to doesHaveViewer role.

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 9 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java` line 116 at r5](https://reviewable.io/reviews/harness/harness-core/42835#-NMA5ydX4XrilXcO77nA-r5-116:-NMY-tJGBiq9OEICG0XL:b-896fjb) ([raw file](https://github.com/harness/harness-core/blob/0ec9a4c5ab71d85d6d811a6e4cdfaa951fb50c14/access-control/service/src/main/java/io/harness/accesscontrol/roleassignments/worker/UserRoleAssignmentRemovalJob.java#L116)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

change to filterAccountsWhereDefaultUserGroupHaveViewerRoleAssignment role.
</blockquote></details>

Done

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 6 of 9 files at r1, 1 of 2 files at r4, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r4, 1 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42835#-:-NMZuL4MDw5oUWP0wYAD:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**